File: src/main/java/net/atos/zac/solr/SolrDeployerService.java
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2024 Lifely
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.zac.solr;

import static net.atos.zac.search.IndexingService.SOLR_CORE;
import static net.atos.zac.solr.FieldType.STRING;
import static net.atos.zac.solr.SolrSchemaUpdateHelper.NAME;
import static net.atos.zac.solr.SolrSchemaUpdateHelper.addField;
import static net.atos.zac.solr.SolrSchemaUpdateHelper.deleteField;

import java.io.IOException;
import java.time.Duration;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.stream.Collectors;

import jakarta.annotation.Resource;
import jakarta.enterprise.concurrent.ManagedExecutorService;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.enterprise.context.Initialized;
import jakarta.enterprise.event.Observes;
import jakarta.enterprise.inject.Instance;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;

import org.apache.commons.lang3.StringUtils;
import org.apache.solr.client.solrj.SolrClient;
import org.apache.solr.client.solrj.SolrServerException;
import org.apache.solr.client.solrj.impl.Http2SolrClient;
import org.apache.solr.client.solrj.request.SolrPing;
import org.apache.solr.client.solrj.request.schema.SchemaRequest;
import org.apache.solr.common.SolrException;
import org.eclipse.microprofile.config.inject.ConfigProperty;

import net.atos.zac.search.IndexingService;
import net.atos.zac.search.model.zoekobject.ZoekObjectType;

@Singleton
public class SolrDeployerService {
    private static final Logger LOG = Logger.getLogger(SolrDeployerService.class.getName());
    private static final String VERSION_FIELD_PREFIX = "schema_version_";
    private static final int SOLR_STATUS_OK = 0;
    private static final int WAIT_FOR_SOLR_SECONDS = 1;

    private String solrUrl;
    private ManagedExecutorService managedExecutor;
    private IndexingService indexingService;
    private SolrClient solrClient;
    private List<SolrSchemaUpdate> schemaUpdates;

    /**
     * Default no-arg constructor, required by Weld.
     */
    public SolrDeployerService() {
    }

    @Inject
    public SolrDeployerService(
            @ConfigProperty(name = "SOLR_URL") final String solrUrl,
            final IndexingService indexingService
    ) {
        this.solrUrl = solrUrl;
        this.indexingService = indexingService;
    }

    @Inject
    public void setSchemaUpdates(final Instance<SolrSchemaUpdate> schemaUpdates) {
        this.schemaUpdates = schemaUpdates.stream()
                .sorted(Comparator.comparingInt(SolrSchemaUpdate::getVersie))
                .toList();
    }

    @Resource
    public void setManagedExecutorService(ManagedExecutorService managedExecutor) {
        this.managedExecutor = managedExecutor;
    }

    public void onStartup(@Observes @Initialized(ApplicationScoped.class) Object event) {
        solrClient = new Http2SolrClient.Builder("%s/solr/%s".formatted(solrUrl, SOLR_CORE)).build();
        waitForSolrAvailability();
        try {
            final int currentVersion = getCurrentVersion();
            LOG.info("Current version of Solr core '%s': %d".formatted(SOLR_CORE, currentVersion));
            if (currentVersion == schemaUpdates.getLast().getVersie()) {
                LOG.info("Solr core '%s' is up to date. No Solr schema migration needed.".formatted(SOLR_CORE));
            } else {
                schemaUpdates.stream()
                        .skip(currentVersion)
                        .forEach(this::apply);

                schemaUpdates.stream()
                        .skip(currentVersion)
                        .flatMap(schemaUpdate -> schemaUpdate.getTeHerindexerenZoekObjectTypes().stream())
                        .collect(Collectors.toSet())
                        .forEach(this::startReindexing);
            }
        } catch (final SolrServerException | IOException e) {
            throw new RuntimeException(e);
        }
    }

    private void waitForSolrAvailability() {
        while (true) {
            try {
                if (new SolrPing().setActionPing().process(solrClient).getStatus() == SOLR_STATUS_OK) {
                    return;
                }
            } catch (final SolrServerException | IOException | SolrException e) {
                LOG.info(() -> "Solr core is not available yet. Exception: %s".formatted(e.getMessage()));
            }
            LOG.info("Waiting for %d seconds for Solr core '%s' to become available...".formatted(WAIT_FOR_SOLR_SECONDS, SOLR_CORE));
            try {
                Thread.sleep(Duration.ofSeconds(WAIT_FOR_SOLR_SECONDS).toMillis());
            } catch (InterruptedException e) {
                LOG.log(
                        Level.WARNING,
                        "Thread was interrupted while waiting for Solr core to become available. Re-interrupting thread.",
                        e
                );
                Thread.currentThread().interrupt();
            }
        }
    }

    private int getCurrentVersion() throws SolrServerException, IOException {
        return new SchemaRequest.Fields().process(solrClient).getFields().stream()
                .map(field -> field.get(NAME).toString())
                .filter(fieldName -> fieldName.startsWith(VERSION_FIELD_PREFIX))
                .findAny()
                .map(versionFieldName -> Integer.valueOf(StringUtils.substringAfter(versionFieldName, VERSION_FIELD_PREFIX)))
                .orElse(0);
    }

    private void apply(final SolrSchemaUpdate schemaUpdate) {
        LOG.info("Updating Solr core '%s' to version: %d".formatted(SOLR_CORE, schemaUpdate.getVersie()));
        try {
            final List<SchemaRequest.Update> schemaUpdates = new LinkedList<>();
            schemaUpdates.addAll(schemaUpdate.getSchemaUpdates());
            schemaUpdates.addAll(updateVersionField(schemaUpdate.getVersie()));
            new SchemaRequest.MultiUpdate(schemaUpdates).process(solrClient);
        } catch (final SolrServerException | IOException e) {
            throw new RuntimeException(e);
        }
    }

    private List<SchemaRequest.Update> updateVersionField(final int version) {
        final List<SchemaRequest.Update> schemaUpdates = new LinkedList<>();
        if (version > 1) {
            schemaUpdates.add(deleteField(VERSION_FIELD_PREFIX + (version - 1)));
        }
        schemaUpdates.add(addField(VERSION_FIELD_PREFIX + version, STRING, false, false));
        return schemaUpdates;
    }

    private void startReindexing(final ZoekObjectType type) {
        managedExecutor.submit(() -> indexingService.reindex(type));
    }
}

File: src/main/kotlin/net/atos/zac/solr/SolrDeployerService.kt
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2024 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package net.atos.zac.solr

import jakarta.annotation.Resource
import jakarta.enterprise.concurrent.ManagedExecutorService
import jakarta.enterprise.context.ApplicationScoped
import jakarta.enterprise.context.Initialized
import jakarta.enterprise.event.Observes
import jakarta.enterprise.inject.Instance
import jakarta.inject.Inject
import jakarta.inject.Singleton
import net.atos.zac.solr.FieldType.STRING
import net.atos.zac.solr.SolrSchemaUpdateHelper.NAME
import net.atos.zac.solr.SolrSchemaUpdateHelper.addField
import net.atos.zac.solr.SolrSchemaUpdateHelper.deleteField
import nl.info.zac.search.IndexingService
import nl.info.zac.search.IndexingService.Companion.SOLR_CORE
import nl.info.zac.search.model.zoekobject.ZoekObjectType
import nl.info.zac.util.AllOpen
import nl.info.zac.util.NoArgConstructor
import org.apache.commons.lang3.StringUtils
import org.apache.solr.client.solrj.SolrClient
import org.apache.solr.client.solrj.SolrServerException
import org.apache.solr.client.solrj.impl.Http2SolrClient
import org.apache.solr.client.solrj.request.SolrPing
import org.apache.solr.client.solrj.request.schema.SchemaRequest
import org.apache.solr.common.SolrException
import org.eclipse.microprofile.config.inject.ConfigProperty
import java.io.IOException
import java.time.Duration
import java.util.logging.Level
import java.util.logging.Logger

@Singleton
@AllOpen
@NoArgConstructor
class SolrDeployerService @Inject constructor(
    @ConfigProperty(name = "SOLR_URL") private val solrUrl: String,
    private val indexingService: IndexingService
) {
    companion object {
        private val LOG = Logger.getLogger(SolrDeployerService::class.java.name)
        private const val VERSION_FIELD_PREFIX = "schema_version_"
        private const val SOLR_STATUS_OK = 0
        private const val INITIAL_WAIT_FOR_SOLR_SECONDS = 1L
        private const val MAX_WAIT_FOR_SOLR_SECONDS = 30L
        private const val WAIT_FOR_SOLR_BACKOFF_MULTIPLIER = 2L
    }

    private lateinit var managedExecutor: ManagedExecutorService
    private lateinit var schemaUpdates: List<SolrSchemaUpdate>
    private lateinit var solrClient: SolrClient

    @Inject
    fun setSchemaUpdates(schemaUpdates: Instance<SolrSchemaUpdate>) {
        this.schemaUpdates = schemaUpdates.sortedBy { it.versie }
    }

    @Resource
    fun setManagedExecutorService(managedExecutor: ManagedExecutorService) {
        this.managedExecutor = managedExecutor
    }

    fun onStartup(@Observes @Initialized(ApplicationScoped::class) @Suppress("UNUSED_PARAMETER") event: Any) {
        solrClient = Http2SolrClient.Builder("$solrUrl/solr/$SOLR_CORE").build()
        waitForSolrAvailability()
        val currentVersion = getCurrentVersion()
        LOG.info { "Current version of Solr core '$SOLR_CORE': $currentVersion" }
        val lastVersion = schemaUpdates.last().versie
        if (currentVersion == lastVersion) {
            LOG.info { "Solr core '$SOLR_CORE' is up to date. No Solr schema migration needed." }
        } else {
            schemaUpdates.drop(currentVersion).forEach(::apply)
            schemaUpdates.drop(currentVersion)
                .flatMap { it.teHerindexerenZoekObjectTypes }
                .toSet()
                .forEach(::startReindexing)
        }
    }

    // Retries with capped exponential backoff instead of a fixed 1-second interval: rootless-networking
    // hiccups between containers (e.g. under Podman) can keep Solr unreachable for well over a minute even
    // though it is otherwise healthy, and hammering the ping every second just adds log noise while that
    // resolves itself. There is no retry limit: Solr is a hard dependency, so waiting indefinitely is correct.
    private fun waitForSolrAvailability() {
        var waitSeconds = INITIAL_WAIT_FOR_SOLR_SECONDS
        while (true) {
            try {
                if (SolrPing().setActionPing().process(solrClient).status == SOLR_STATUS_OK) {
                    return
                }
            } catch (exception: SolrServerException) {
                LOG.info { "Solr core is not available yet. Exception: ${exception.message}" }
            } catch (exception: IOException) {
                LOG.info { "Solr core is not available yet. Exception: ${exception.message}" }
            } catch (exception: SolrException) {
                LOG.info { "Solr core is not available yet. Exception: ${exception.message}" }
            }
            LOG.info { "Waiting for $waitSeconds seconds for Solr core '$SOLR_CORE' to become available..." }
            try {
                Thread.sleep(Duration.ofSeconds(waitSeconds).toMillis())
            } catch (exception: InterruptedException) {
                LOG.log(
                    Level.WARNING,
                    "Thread was interrupted while waiting for Solr core to become available. Re-interrupting thread.",
                    exception
                )
                Thread.currentThread().interrupt()
            }
            waitSeconds = (waitSeconds * WAIT_FOR_SOLR_BACKOFF_MULTIPLIER).coerceAtMost(MAX_WAIT_FOR_SOLR_SECONDS)
        }
    }

    private fun getCurrentVersion(): Int =
        SchemaRequest.Fields().process(solrClient).fields
            .map { it[NAME].toString() }
            .firstOrNull { it.startsWith(VERSION_FIELD_PREFIX) }
            ?.let { StringUtils.substringAfter(it, VERSION_FIELD_PREFIX).toInt() }
            ?: 0

    private fun apply(schemaUpdate: SolrSchemaUpdate) {
        LOG.info { "Updating Solr core '$SOLR_CORE' to version: ${schemaUpdate.versie}" }
        val updates = schemaUpdate.schemaUpdates + updateVersionField(schemaUpdate.versie)
        SchemaRequest.MultiUpdate(updates).process(solrClient)
    }

    private fun updateVersionField(version: Int): List<SchemaRequest.Update> {
        val updates = mutableListOf<SchemaRequest.Update>()
        if (version > 1) {
            updates.add(deleteField(VERSION_FIELD_PREFIX + (version - 1)))
        }
        updates.add(addField(VERSION_FIELD_PREFIX + version, STRING, false, false))
        return updates
    }

    private fun startReindexing(type: ZoekObjectType) {
        managedExecutor.submit { indexingService.reindex(type) }
    }
}
